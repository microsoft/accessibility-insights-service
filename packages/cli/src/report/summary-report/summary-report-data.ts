// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

export interface ViolationCountMap {
    [key: string]: number;
}

export interface UrlToReportMap {
    [key: string]: string;
}

export interface SummaryReportData {
    violationCountByRuleMap: ViolationCountMap;
    failedUrlToReportMap: UrlToReportMap;
    passedUrlToReportMap: UrlToReportMap;
    unScannableUrls: string[];
}
