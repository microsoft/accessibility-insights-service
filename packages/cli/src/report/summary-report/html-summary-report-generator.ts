// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as escapeHtml from 'escape-html';
import { injectable } from 'inversify';
import { SummaryReportData, UrlToReportMap } from './summary-report-data';
import { SummaryReportGenerator } from './summary-report-generator';

// tslint:disable-next-line: no-var-requires no-require-imports no-unsafe-any
const pretty: (html: string) => string = require('pretty');

type HtmlSummaryLink = {
    link: string;
    fileName: string;
};

type HtmlSummaryData = {
    passedLinks: HtmlSummaryLink[];
    failedLinks: HtmlSummaryLink[];
};

@injectable()
export class HtmlSummaryReportGenerator implements SummaryReportGenerator {
    public generateReport(summaryReportData: SummaryReportData): string {
        const htmlSummaryData: HtmlSummaryData = {
            failedLinks: [],
            passedLinks: [],
        };

        htmlSummaryData.failedLinks = this.getHtmlLinkData(summaryReportData.failedUrlToReportMap);
        htmlSummaryData.passedLinks = this.getHtmlLinkData(summaryReportData.passedUrlToReportMap);

        return pretty(`
        <!DOCTYPE html>
        <html lang='en'>
            <head>
                <meta charset="UTF-8">
                <title>Accessibility Insights Summary Report</title>
            </head>
            <body>
                ${this.getHtmlBody(htmlSummaryData)}
            </body>
        </html>
        `);
    }

    private getHtmlBody(htmlSummaryData: HtmlSummaryData): string {
        const totalUrlsScannedInfo = `
        <b>
            Total Urls Scanned - ${htmlSummaryData.failedLinks.length + htmlSummaryData.passedLinks.length}
        </b>
        `;
        if (htmlSummaryData.failedLinks.length === 0 && htmlSummaryData.passedLinks.length === 0) {
            return totalUrlsScannedInfo;
        }

        return `
        ${totalUrlsScannedInfo}
        <ul>
            ${this.getFailedLinksHtml(htmlSummaryData)}
            ${this.getPassedLinksHtml(htmlSummaryData)}
        </ul>`;
    }

    private getFailedLinksHtml(htmlSummaryData: HtmlSummaryData): string {
        if (htmlSummaryData.failedLinks.length === 0) {
            return '';
        }

        return this.getLinksSection('Failed Urls', htmlSummaryData.failedLinks);
    }

    private getPassedLinksHtml(htmlSummaryData: HtmlSummaryData): string {
        if (htmlSummaryData.passedLinks.length === 0) {
            return '';
        }

        return this.getLinksSection('Passed Urls', htmlSummaryData.passedLinks);
    }

    private getLinksSection(sectionName: string, links: HtmlSummaryLink[]): string {
        return `
        <li>
            ${sectionName}
            <ul>
                ${this.getHtmlLinks(links)}
            </ul>
        </li>`;
    }

    private getHtmlLinks(links: HtmlSummaryLink[]): string {
        const htmlLinks: string[] = links.map((item) => {
            return `
            <li>
                <a href='${item.link}'>${escapeHtml(item.fileName)}</a>
            </li>`;
        });

        return htmlLinks.join('\n');
    }

    private getHtmlLinkData(urlMap: UrlToReportMap): HtmlSummaryLink[] {
        const links: HtmlSummaryLink[] = [];

        for (const url of Object.keys(urlMap)) {
            links.push({
                fileName: url,
                link: urlMap[url],
            });
        }

        return links;
    }
}
