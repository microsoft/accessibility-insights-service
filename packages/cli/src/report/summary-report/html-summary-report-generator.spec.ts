// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import 'reflect-metadata';

import { HtmlSummaryReportGenerator } from './html-summary-report-generator';
import { SummaryReportData } from './summary-report-data';

/* eslint-disable @typescript-eslint/consistent-type-assertions */

describe(HtmlSummaryReportGenerator, () => {
    let testSubject: HtmlSummaryReportGenerator;

    beforeEach(() => {
        testSubject = new HtmlSummaryReportGenerator();
    });

    test.each([
        [
            'has no scanned urls',
            {
                failedUrlToReportMap: {},
                passedUrlToReportMap: {},
                violationCountByRuleMap: {},
                unscannableUrls: {},
            } as SummaryReportData,
        ],
        [
            'has failed urls only',
            {
                failedUrlToReportMap: {
                    'https://www.failedUrl1.com': './failed-url1.html',
                    'https://www.failedUrl2.com': './failed-url2.html',
                },
                passedUrlToReportMap: {},
                violationCountByRuleMap: { rule1: 3 },
                unscannableUrls: {},
            } as SummaryReportData,
        ],
        [
            'has passed urls only',
            {
                failedUrlToReportMap: {},
                passedUrlToReportMap: {
                    'https://www.passedUrl1.com': './passed-url1.html',
                    'https://www.passedUrl2.com': './passed-url2.html',
                },
                violationCountByRuleMap: {},
                unscannableUrls: {},
            } as SummaryReportData,
        ],
        [
            'has both failed & passed urls',
            {
                failedUrlToReportMap: {
                    'https://www.failedUrl1.com': './failed-url1.html',
                    'https://www.failedUrl2.com': './failed-url2.html',
                },
                passedUrlToReportMap: {
                    'https://www.passedUrl1.com': './passed-url1.html',
                    'https://www.passedUrl2.com': './passed-url2.html',
                },
                violationCountByRuleMap: {
                    rule1: 3,
                },
                unscannableUrls: {},
            } as SummaryReportData,
        ],
        [
            'has unScannable urls only',
            {
                failedUrlToReportMap: {},
                passedUrlToReportMap: {},
                violationCountByRuleMap: {},
                unscannableUrls: {
                    'https://www.unscannableUrl1.com': './unscannableUrl1',
                },
            } as SummaryReportData,
        ],
    ])('logs violation summary when %s', async (testCaseName, testCase) => {
        expect(testSubject.generateReport(testCase)).toMatchSnapshot();
    });
});
