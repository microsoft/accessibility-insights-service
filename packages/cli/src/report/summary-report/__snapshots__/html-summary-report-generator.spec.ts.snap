// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HtmlSummaryReportGenerator logs violation summary when has both failed & passed urls 1`] = `
"<!DOCTYPE html>
<html lang='en'>

  <head>
    <meta charset=\\"UTF-8\\">
    <title>Accessibility Insights Summary Report</title>
  </head>

  <body>


    <b>
      Total Urls Scanned - 4
    </b>

    <ul>

      <li>
        Failed Urls
        <ul>

          <li>
            <a href='./failed-url1.html'>https://www.failedUrl1.com</a>
          </li>

          <li>
            <a href='./failed-url2.html'>https://www.failedUrl2.com</a>
          </li>
        </ul>
      </li>

      <li>
        Passed Urls
        <ul>

          <li>
            <a href='./passed-url1.html'>https://www.passedUrl1.com</a>
          </li>

          <li>
            <a href='./passed-url2.html'>https://www.passedUrl2.com</a>
          </li>
        </ul>
      </li>
    </ul>
  </body>

</html>"
`;

exports[`HtmlSummaryReportGenerator logs violation summary when has failed urls only 1`] = `
"<!DOCTYPE html>
<html lang='en'>

  <head>
    <meta charset=\\"UTF-8\\">
    <title>Accessibility Insights Summary Report</title>
  </head>

  <body>


    <b>
      Total Urls Scanned - 2
    </b>

    <ul>

      <li>
        Failed Urls
        <ul>

          <li>
            <a href='./failed-url1.html'>https://www.failedUrl1.com</a>
          </li>

          <li>
            <a href='./failed-url2.html'>https://www.failedUrl2.com</a>
          </li>
        </ul>
      </li>

    </ul>
  </body>

</html>"
`;

exports[`HtmlSummaryReportGenerator logs violation summary when has no scanned urls 1`] = `
"<!DOCTYPE html>
<html lang='en'>

  <head>
    <meta charset=\\"UTF-8\\">
    <title>Accessibility Insights Summary Report</title>
  </head>

  <body>

    <b>
      Total Urls Scanned - 0
    </b>

  </body>

</html>"
`;

exports[`HtmlSummaryReportGenerator logs violation summary when has passed urls only 1`] = `
"<!DOCTYPE html>
<html lang='en'>

  <head>
    <meta charset=\\"UTF-8\\">
    <title>Accessibility Insights Summary Report</title>
  </head>

  <body>


    <b>
      Total Urls Scanned - 2
    </b>

    <ul>


      <li>
        Passed Urls
        <ul>

          <li>
            <a href='./passed-url1.html'>https://www.passedUrl1.com</a>
          </li>

          <li>
            <a href='./passed-url2.html'>https://www.passedUrl2.com</a>
          </li>
        </ul>
      </li>
    </ul>
  </body>

</html>"
`;
